### PR TITLE
SONARJAVA-2899 Loggers configuration is a security hotspot

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S4792_java.html [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+<p>This rule supports the following libraries: Log4J, <code>java.util.logging</code> and Logback</p>
+<pre>
+// === Log4J 2 ===
+import org.apache.logging.log4j.core.config.builder.api.ConfigurationBuilderFactory;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.core.*;
+import org.apache.logging.log4j.core.config.*;
+
+abstract class CustomConfigFactory extends ConfigurationFactory {
+    // ...
```
Comment (christophe-zurn-sonarsource) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S4792_java.html]: RSPEC should describe this part of the code as `Questionable`, and why it is so

File: java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      MethodMatcher.create().typeDefinition(LOG4J_CONFIGURATOR).name(SET_LEVEL).withAnyParameters(),
+      MethodMatcher.create().typeDefinition(LOG4J_CONFIGURATOR).name("setRootLevel").withAnyParameters(),
+      MethodMatcher.create().typeDefinition("org.apache.logging.log4j.core.config.Configuration").name("addAppender").withAnyParameters(),
+      MethodMatcher.create().typeDefinition("org.apache.logging.log4j.core.LoggerContext").name("setConfigLocation").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(LOG4J_CONFIGURATION_SOURCE).name("<init>").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(LOG4J_CONFIGURATION_SOURCE).name("fromResource").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(LOG4J_CONFIGURATION_SOURCE).name("fromUri").withAnyParameters(),
+      MethodMatcher.create().typeDefinition("java.util.logging.LogManager").name("readConfiguration").withAnyParameters(),
+      MethodMatcher.create().typeDefinition("java.util.logging.Logger").name(SET_LEVEL).withAnyParameters(),
+      MethodMatcher.create().typeDefinition("java.util.logging.Logger").name("addHandler").withAnyParameters(),
```
Comment (christophe-zurn-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java]: `java.util.logging.LoggingMXBean#setLoggerLevel(String loggerName, String levelName)` could also be considered to raise an issue

File: java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return Arrays.asList(
+      MethodMatcher.create().typeDefinition("org.apache.logging.log4j.core.config.builder.api.ConfigurationBuilderFactory").name("newConfigurationBuilder").withoutParameter(),
+      MethodMatcher.create().typeDefinition(LOG4J_CONFIGURATOR).name("setAllLevels").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(LOG4J_CONFIGURATOR).name(SET_LEVEL).withAnyParameters(),
+      MethodMatcher.create().typeDefinition(LOG4J_CONFIGURATOR).name("setRootLevel").withAnyParameters(),
+      MethodMatcher.create().typeDefinition("org.apache.logging.log4j.core.config.Configuration").name("addAppender").withAnyParameters(),
```
Comment (christophe-zurn-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java]: Is there a reason the `addAppender` and `setLevel` methods for `org.apache.logging.log4j.core.config.LoggerConfig` are not included as violations for log4j?

File: java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      MethodMatcher.create().typeDefinition("ch.qos.logback.classic.Logger").name(SET_LEVEL).withAnyParameters(),
+      MethodMatcher.create().typeDefinition("ch.qos.logback.classic.joran.JoranConfigurator").name("<init>").withAnyParameters(),
+      MethodMatcher.create().typeDefinition("java.lang.System").name("setProperty").parameters("java.lang.String", "java.lang.String")
+      );
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.CLASS)) {
+      checkConfigurationFactoryExtension(((ClassTree) tree));
```
Comment (christophe-zurn-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java]: We could also raise issues for the `logback` library on classes extending `ch.qos.logback.classic.spi.Configurator`

**Comment by benzonico** [comment]
Regarding XML files, this is for me an XML rule and should be written as such then.

**Comment by nicolas-harraudeau-sonarsource** [comment]
@christophe-zurn-sonarsource I confirm that the configuration files are for now out of scope. I updated the comment to keep the handling of config as a memo instead of a task for this MMF.

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S4792_java.html (hunk 1 above)
Comment (nicolas-harraudeau-sonarsource) [reply java-checks/src/main/resources/org/sonar/l10n/java/rules/squid/S4792_java.html]: Fixed, I removed it by mistake while making the example executable.

File: java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java (hunk 2 above)
Comment (nicolas-harraudeau-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java]: I'll keep this in mind for an improvement of the rule later. Other interfaces could also be included like `PlatformLoggingMXBean` but we can't cover every logging API right away.

File: java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java (hunk 3 above)
Comment (nicolas-harraudeau-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java]: Good catch, I'll update the example on Jira with those two methods.

**Comment by nicolas-harraudeau-sonarsource** [comment]
@benzonico RSPEC updated. I added to the list of Questionable the following: `addAppender` and `setLevel` from `org.apache.logging.log4j.core.config.LoggerConfig`

File: java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java (hunk 2 above)
Comment (christophe-zurn-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java]: Fine by me :)

File: java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java (hunk 4 above)
Comment (christophe-zurn-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/security/LogConfigurationCheck.java]: Ok :+1:

**Pull request merged**